crates/sparrow-merge/src/merge/heterogeneous_merge.rs

```diff
@@ -0,0 +1,276 @@
+use std::collections::BinaryHeap;
+use std::sync::Arc;
+
+use anyhow::Context;
+use arrow_array::{ArrayRef, RecordBatch};
+use arrow_schema::{DataType, SchemaRef};
+use itertools::izip;
+use sparrow_batch::Batch;
+
+use crate::{gather::Gatherer, spread::Spread};
+
+use super::{binary_merge, BinaryMergeResult};
+
+/// Manages the merge of two heterogeneous inputs.
+pub struct HeterogeneousMerge {
+    result_type: DataType,
+    spread_l: Spread,
+    spread_r: Spread,
+
+    /// Gathers batches from both sides and produces [GatheredBatches]
+    /// up to a valid watermark.
+    gatherer: Gatherer,
+}
+
+#[derive(derive_more::Display, Debug)]
+pub enum Error {
```

Comment:
For now, if we think this will always be just one error, we could simplify to:

```
#[display(fmt = "internal error merging: {}", _0)]
pub struct Error(&'static str);
```